### DISCUSSION TITLE
Performance / Delay Issue with Wezterm

### DISCUSSION THREAD

**Discussion by laxman20** [mrjones2014/smart-splits.nvim#133] [General]
I'm noticing a delay when switching panes when using Wezterm.

Moving from Neovim pane to Neovim pane is quick.
Moving from Neovim pane to Wezterm pane has a noticeable delay.
Wezterm to Wezterm and Wezterm to Neovim are quick

Wezterm version: `20230712-072601-f4abf8fd`

Wezterm config:

```
local w = require('wezterm')
local config = {}

if w.config_builder then
	config = w.config_builder()
end

config.window_padding = {
	left = 0,
	right = 0,
	top = 0,
	bottom = 0,
}

-- config.color_scheme = 'OneDark (base16)'
config.color_scheme = 'nightfox'


local function is_vim(pane)
	return pane:get_user_vars().IS_NVIM == 'true'
end

local direction_keys = {
	Left = 'h',
	Down = 'j',
	Up = 'k',
	Right = 'l',
	-- reverse lookup
	h = 'Left',
	j = 'Down',
	k = 'Up',
	l = 'Right',
}

local function split_nav(resize_or_move, key)
	return {
		key = key,
		mods = resize_or_move == 'resize' and 'META' or 'CTRL',
		action = w.action_callback(function(win, pane)
			if is_vim(pane) then
				-- pass the keys through to vim/nvim
				win:perform_action({
						SendKey = { key = key, mods = resize_or_move == 'resize' and 'META' or 'CTRL' },
					}, pane)
			else
				if resize_or_move == 'resize' then
					win:perform_action({ AdjustPaneSize = { direction_keys[key], 3 } }, pane)
				else
					win:perform_action({ ActivatePaneDirection = direction_keys[key] }, pane)
				end
			end
		end),
	}
end

config.keys = {
	-- move between split panes
	split_nav('move', 'h'),
	split_nav('move', 'j'),
	split_nav('move', 'k'),
	split_nav('move', 'l'),
	-- resize panes
	split_nav('resize', 'h'),
	split_nav('resize', 'j'),
	split_nav('resize', 'k'),
	split_nav('resize', 'l'),
	-- splits
	{ key = '|', mods = 'SUPER', action = w.action.SplitHorizontal { domain = 'CurrentPaneDomain' }},
	{ key = '_', mods = 'SUPER', action = w.action.SplitVertical { domain = 'CurrentPaneDomain' }},
	{ key = '|', mods = 'SHIFT|SUPER', action = w.action.SplitHorizontal { domain = 'CurrentPaneDomain' }},
	{ key = '_', mods = 'SHIFT|SUPER', action = w.action.SplitVertical { domain = 'CurrentPaneDomain' }},
	-- remove Al+Enter binding
	{ key = 'Enter', mods = 'ALT', action = w.action.DisableDefaultAssignment },
}

return config
```

Neovim version: v0.9.2

Smart-Splits version: 1.2.4
Smart-Splits config (lazy config):

```
return {
	"mrjones2014/smart-splits.nvim",
	lazy = false,
	config = function()
		vim.keymap.set('n', '<C-h>', require('smart-splits').move_cursor_left)
		vim.keymap.set('n', '<C-j>', require('smart-splits').move_cursor_down)
		vim.keymap.set('n', '<C-k>', require('smart-splits').move_cursor_up)
		vim.keymap.set('n', '<C-l>', require('smart-splits').move_cursor_right)


		vim.keymap.set('n', '<A-h>', require('smart-splits').resize_left)
		vim.keymap.set('n', '<A-j>', require('smart-splits').resize_down)
		vim.keymap.set('n', '<A-k>', require('smart-splits').resize_up)
		vim.keymap.set('n', '<A-l>', require('smart-splits').resize_right)
	end
}
```

**Comment by mrjones2014** [comment]
Hmm... This all looks like exactly what I have. Are your dotfiles in a public repo? If so I can take your full Wezterm + Neovim configs for a spin locally and see if I can narrow it down from there.

**Reply by laxman20** [reply]
Should be public now https://github.com/laxman20/dotfiles

**Reply by mrjones2014** [reply]
Cool. I'll try them out when I have time and see if I can find anything that might be causing that.

**Reply by laxman20** [reply]
So I had some time to look into this myself. Loaded different versions of the plugin and found that the issue starts with v1.2.2. On v1.2.1 everything is snappy.

Unfortunately, on v1.2.1, my up and down motions stop working (this is resolved in 1.2.2).

**Reply by laxman20** [reply]
Looked into it further. It's the addition of smart-splits.vim that's causing it https://github.com/mrjones2014/smart-splits.nvim/compare/v1.2.1...v1.2.2 

Issue goes away when I comment out the if block at the bottom

**Reply by mrjones2014** [reply]
Hmm, that doesn't really sound right to me. That `if` block at the bottom of `smart-splits.vim` only runs once at startup, and once right before exiting nvim. It shouldn't have any effect on normal movement between panes 🤔

**Reply by laxman20** [reply]
Ya you're right. Commenting out `call s:write(s:format_var("true"))` does get rid of the delay though.

Someone on Reddit let me know that it's happening for them on Windows, but not Linux.
I'm able to reproduce it on Windows and Mac.

**Reply by laxman20** [reply]
Setting verbose=9 and doing a pane switch makes a lot of wezterm cli calls

```
Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'list' '--format' 'json'"

Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'list' '--format' 'json'"
Press ENTER or type command to continue

Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'list' '--format' 'json'"
Press ENTER or type command to continue

Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'list' '--format' 'json'"
Press ENTER or type command to continue

Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'list' '--format' 'json'"
Press ENTER or type command to continue

Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'activate-pane-direction' 'Right'"
Press ENTER or type command to continue

Executing command: "'/opt/homebrew/bin/wezterm' 'cli' 'list' '--format' 'json'"
Press ENTER or type command to continue
```

I don't see the same on v1.2.1

**Reply by mrjones2014** [reply]
Are those logs for a single pane switch? If so it looks like it's doing `wezterm cli list --format json` way too many times for some reason 🤔

**Reply by laxman20** [reply]
Yup that’s for a single switch from left pane to right.

**Reply by mrjones2014** [reply]
Alright that's a good clue then. It does have to get current pane info a few times to double check that things are working properly, but that looks like too many to me. I'll open an issue to investigate further.

**Comment by NullSeile** [comment]
Hello! I've encountered the same problem with the plugin. Exactly equal to what @laxman20 explained, the same version caused the issue, and 1.2.1 works fine. Has There been any updates?

**Comment by mrjones2014** [comment]
Basically the issue is likely with your shell's startup performance. It shouldn't affect it, but it does for some reason. Details in #261
